airbyte-workers/src/main/java/io/airbyte/workers/helper/FailureHelper.java

```diff
@@ -34,18 +35,62 @@ public static FailureReason genericFailure(final Throwable t, final Long jobId,
         .withMetadata(jobAndAttemptMetadata(jobId, attemptNumber));
   }
 
+  public static FailureReason genericFailure(final AirbyteTraceMessage m, final Long jobId, final Integer attemptNumber) {
+    return new FailureReason()
+        .withInternalMessage(m.getError().getInternalMessage())
+        .withStacktrace(m.getError().getStackTrace())
+        .withTimestamp(m.getEmittedAt().longValue())
+        .withMetadata(traceMessageMetadata(jobId, attemptNumber));
+  }
+
   public static FailureReason sourceFailure(final Throwable t, final Long jobId, final Integer attemptNumber) {
     return genericFailure(t, jobId, attemptNumber)
         .withFailureOrigin(FailureOrigin.SOURCE)
         .withExternalMessage("Something went wrong within the source connector");
   }
 
+  public static FailureReason sourceFailure(final AirbyteTraceMessage m, final Long jobId, final Integer attemptNumber) {
+    return genericFailure(m, jobId, attemptNumber)
+        .withFailureOrigin(FailureOrigin.SOURCE)
+        .withExternalMessage(m.getError().getMessage());
+  }
+
   public static FailureReason destinationFailure(final Throwable t, final Long jobId, final Integer attemptNumber) {
     return genericFailure(t, jobId, attemptNumber)
         .withFailureOrigin(FailureOrigin.DESTINATION)
         .withExternalMessage("Something went wrong within the destination connector");
   }
 
+  public static FailureReason destinationFailure(final AirbyteTraceMessage m, final Long jobId, final Integer attemptNumber) {
+    return genericFailure(m, jobId, attemptNumber)
+        .withFailureOrigin(FailureOrigin.DESTINATION)
+        .withExternalMessage(m.getError().getMessage());
+  }
+
+  public static FailureReason errorTraceMessageFailure(final AirbyteTraceMessage sourceMessage,
```

Comment:
Most of the logic in this method is picking between the source and destination message based on which one was emitted first. I think this logic would make more sense to put in the AirbyteMessageTracker rather than in the FailureHelper, since this class is really just meant to help with constructing FailureReasons.

If this is moved into AirbyteMessageTracker, we probably no longer need to call `getFirstSourceErrorTraceMessage` and `getFirstDestinationErrorTraceMessage` in the replication worker - it can just retrieve the one "first error trace message" from the message tracker and add it to `failures` if present